TimeSeries: Add support for negative Y and constant transform

dprokop opened the pull request:
Closes https://github.com/grafana/grafana/discussions/38202

This PR brings highly anticipated **Negative Y transform** to Time series panel, as well as **Constant transform**.

TODO:
- [x] Old graph -> time series migration
- [x] Documentation 

[Test dashboard.json](https://gist.github.com/dprokop/639f9f830ddf3d0591bebd9a2225058a)

Configuration UI (available **only** as an override):

![image](https://user-images.githubusercontent.com/2376619/152164843-37acdb64-9bb4-4aa9-a59a-fca0b21de60c.png)

UI indicators about series being transformed:

|  Tooltip | Legend  |
|---|---|
|![image](https://user-images.githubusercontent.com/2376619/152165085-51a18525-28c5-4389-abca-1c0c0f5bce2a.png)|![image](https://user-images.githubusercontent.com/2376619/152165137-0da3bb21-1412-4c9c-9ffd-a6173406d03b.png)<br/>![image](https://user-images.githubusercontent.com/2376619/152165247-86f586af-6355-4671-9408-7a15b1334ec1.png)|

Old graph vs timeseries basic comparison:

![image](https://user-images.githubusercontent.com/2376619/152343822-c43d6170-a32c-49d2-b1df-06b825f5facd.png)

ryantxu committed:
add negative y config

ryantxu committed:
Merge remote-tracking branch 'origin/main' into neg-y

dprokop committed:
Merge branch 'main' into neg-y

dprokop committed:
Handle negative y and constant transform in Timeseries panel

dprokop commented on public/app/plugins/panel/timeseries/config.ts:
```diff
@@ -181,6 +182,29 @@ export function getGraphFieldConfig(cfg: GraphFieldConfig): SetFieldConfigOption
+
+      builder.addSelect({
+        category: categoryStyles,
+        name: 'Transform',
```
Not sure this name is clear enough, but can't think of anything else. Any suggestions are highly appreciated

dprokop committed:
Typechecks

dprokop committed:
Add migration from old graph panel

dprokop committed:
Docs update

ryantxu commented on public/app/plugins/panel/state-timeline/utils.ts:
```diff
@@ -325,8 +325,8 @@ export function mergeThresholdValues(field: Field, theme: GrafanaTheme2): Field
   for (let i = 0; i < items.length; i++) {
-    thresholdToText.set(thresholds.steps[i], items[i].label);
-    textToColor.set(items[i].label, items[i].color!);
+    thresholdToText.set(thresholds.steps[i], items[i].name);
+    textToColor.set(items[i].name, items[i].color!);
```
:thinking: looks like name is used here.    So we can not just remove it

ryantxu reviewed (commented):
Looks good -- I am not sure about duplicate name, label, and key on the legend.   But also not sure I see a quick way to avoid it

ryantxu commented:
If this is merged, is everything on the parity list supported?

dprokop commented:
> Looks good -- I am not sure about duplicate name, label, and key on the legend. But also not sure I see a quick way to avoid it

Don't like it neither - thrown quickly to get a quick feedback on the feat. Will revisit tomorrow morning :)

dprokop commented on packages/grafana-data/src/field/overrides/processors.ts:
```diff
@@ -74,6 +74,8 @@ export const valueMappingsOverrideProcessor = (
 export interface SelectFieldConfigSettings<T> {
   allowCustomValue?: boolean;
 
+  isClearable?: boolean;
```
Guess who added it.... ;)

simPod reviewed (changes requested):
![image](https://user-images.githubusercontent.com/327717/152238302-976d23bc-30bc-4fc8-9ba5-c422c734792a.png)

Do we need the `(neg Y)` there? I often have like 40 series and I think that it will be a noise. Series name is IMO sufficient.

nward commented:
Hi @dprokop this is great work thanks for getting in to this.

The example dashboard has negative data in it - I think (at least for me!) the use case is to use positive values, but have it show below 0. This is particularly important for my use case as we select data by "highest" - if we have to translate it to negative then "highest" is no long really what we want.

One use case for this visualisation is for a network link, showing traffic either above or below 0 depending on the direction (rx vs. tx).
It is also the case that we would want to have multiple network links on the same chart, with their rx values stacked, and tx values stacked - with rx showing above 0 and tx showing below 0.

A couple of things I'm interested in / think you should include in your testing:
- 2 Series of positive value data, one positive (normal), and one negative-y (which should render below 0)
- 4 series of positive value data, 2 positive (normal), 2 negative-Y. The positive and negative stacks should be distinct, and should result in the stacked "negative-y" value rendering further below 0 than if it was not stacked. Perhaps stacks should automatically be separated based on negative/positive Y? I'm not sure of an example of when you'd want the negative-Y to be on the same stack as positive-Y, so feels like this should just be automatic - i.e. if you have a stack "a" on a positive-y value it should automatically be distinct from any series with stack "a" with negative-y.

nward commented:
> Do we need the `(neg Y)` there? I often have like 40 series and I think that it will be a noise. Series name is IMO sufficient.

I agree with this - I don't think this is valuable for what people use this feature for. Perhaps showing this in the label could be a default disabled option?

dprokop committed:
Revert "Add migration from old graph panel"

This reverts commit 33b5a90b66bf0ea7fd294aeb8da4774ccdff1a0a.

dprokop committed:
Revert VizLegendItem changes

dprokop committed:
Automatically separate positive and negative stacks within a group

dprokop commented:
@nward @simPod  - Thanks a bunch for feedback. As of now, I have removed the transform indicators from the tooltip and the legend. 

@nward  - Me and @leeoniya have been discussing separate neg/pos y stacks a while ago. I have implemented it in the last commit. The gist is that within a single stacking group it will create a pseudo-group for the series that has negative Y transform applied and will result in a separate stack on the neg y-axis. See the screenshot in the PR description. I have also updated the test dashboard with a few more cases - feel free to test it out against this branch.

dprokop commented:
> Looks good -- I am not sure about duplicate name, label, and key on the legend. But also not sure I see a quick way to avoid it

@ryantxu - I have reverted this change as it was only to support to transformation indicator in legend/tooltip - according to the feedback above it seems like it might be a useless thing. We can revisit it in the future

dprokop commented on packages/grafana-ui/src/components/VizLegend/VizLegend.story.tsx:
```diff
@@ -158,6 +158,7 @@ function generateLegendItems(
   return [...new Array(numberOfSeries)].map((item, i) => {
     return {
       label: `${alphabet[i].toUpperCase()}-series`,
+      name: `${alphabet[i].toUpperCase()}-series`,
```
```suggestion
```

dprokop committed:
Update packages/grafana-ui/src/components/VizLegend/VizLegend.story.tsx

dprokop committed:
Remove SeriesLabel component

achatterjee-grafana commented on docs/sources/visualizations/time-series/_index.md:
```diff
@@ -35,6 +35,13 @@ Use these options to choose how to display your time series data.
 
+### Transform
+
+Use this option to transform series values without affecting values shown in tooltip, context menu and legend.
```
```suggestion
Use this option to transform the series values without affecting the values shown in the tooltip, context menu, and legend.
```

achatterjee-grafana commented on docs/sources/visualizations/time-series/_index.md:
```diff
@@ -35,6 +35,13 @@ Use these options to choose how to display your time series data.
+
+Use this option to transform series values without affecting values shown in tooltip, context menu and legend.
+
+- **Negative Y transform -** Flip the results to negative values on the y axis.
```
```suggestion
- **Negative Y transform -** Flip the results to negative values on the Y axis.
```

achatterjee-grafana reviewed (commented):
Added doc review.

dprokop committed:
Update docs/sources/visualizations/time-series/_index.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

dprokop committed:
Update docs/sources/visualizations/time-series/_index.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

dprokop committed:
Leftover reverts

dprokop committed:
Don't crate bands (for now0 for negative -y stack

dprokop committed:
Add docs note about transform being only available as an override

dprokop committed:
Merge branch 'neg-y' of github.com:grafana/grafana into neg-y

dprokop commented:
> If this is merged, is everything on the parity list supported?

The last thing missing after this is regions support (https://github.com/grafana/grafana/issues/35486)

nward commented:
> @nward @simPod - Thanks a bunch for feedback. As of now, I have removed the transform indicators from the tooltip and the legend.
> 
> @nward - Me and @leeoniya have been discussing separate neg/pos y stacks a while ago. I have implemented it in the last commit. The gist is that within a single stacking group it will create a pseudo-group for the series that has negative Y transform applied and will result in a separate stack on the neg y-axis. See the screenshot in the PR description. I have also updated the test dashboard with a few more cases - feel free to test it out against this branch.

Nice work! That's really excellent. I'll see about getting this branch deployed somewhere it can talk to my real Influxdb data and see how it goes and get back to you.

leeoniya commented:
> If this is merged, is everything on the parity list supported?
> @ryantxu The last thing missing after this is regions support (https://github.com/grafana/grafana/issues/35486)

and maybe https://github.com/grafana/grafana/discussions/38299?

dprokop committed:
Merge branch 'main' into neg-y

dprokop committed:
Merge branch 'main' into neg-y

dprokop committed:
Fill negative bands in reversed direction

zoltanbedi reviewed (approved):
Tried it out. LGTM!

dprokop committed:
Merge branch 'main' into neg-y

IR31108 commented:
I'm in 9.2.4 and I'm unable to do this for a time series with only one data line. It behaves the same as multiplying by -1 which is not helpful in my case. Note that the axis is in the negatives which would be misleading. Any idea why or how to fix it?

![image](https://user-images.githubusercontent.com/100949749/228913151-72f5b68d-9311-42b1-b9c3-eec00cf295d5.png)
